// @strapi/content-releases/dist/_chunks/App-f2cafd81.js
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("react/jsx-runtime");
const reactRouterDom = require("react-router-dom");
const index = require("./index-66d129ac.js");
const React = require("react");
const designSystem = require("@strapi/design-system");
const v2 = require("@strapi/design-system/v2");
const helperPlugin = require("@strapi/helper-plugin");
const icons = require("@strapi/icons");
const reactIntl = require("react-intl");
const styled = require("styled-components");
const formik = require("formik");
const yup = require("yup");
require("@reduxjs/toolkit/query");
require("axios");
require("@reduxjs/toolkit/query/react");
const _interopDefault = (e) => e && e.__esModule ? e : { default: e };
function _interopNamespace(e) {
  if (e && e.__esModule)
    return e;
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const React__namespace = /* @__PURE__ */ _interopNamespace(React);
const styled__default = /* @__PURE__ */ _interopDefault(styled);
const yup__namespace = /* @__PURE__ */ _interopNamespace(yup);
const RELEASE_SCHEMA = yup__namespace.object().shape({
  name: yup__namespace.string().trim().required()
}).required().noUnknown();
const ReleaseModal = ({
  handleClose,
  handleSubmit,
  initialValues,
  isLoading = false
}) => {
  const { formatMessage } = reactIntl.useIntl();
  return /* @__PURE__ */ jsxRuntime.jsxs(designSystem.ModalLayout, { onClose: handleClose, labelledBy: "title", children: [
    /* @__PURE__ */ jsxRuntime.jsx(designSystem.ModalHeader, { children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { id: "title", fontWeight: "bold", textColor: "neutral800", children: formatMessage({
      id: "content-releases.modal.add-release-title",
      defaultMessage: "New release"
    }) }) }),
    /* @__PURE__ */ jsxRuntime.jsx(
      formik.Formik,
      {
        validateOnChange: false,
        onSubmit: handleSubmit,
        initialValues,
        validationSchema: RELEASE_SCHEMA,
        children: ({ values, errors, handleChange }) => /* @__PURE__ */ jsxRuntime.jsxs(formik.Form, { children: [
          /* @__PURE__ */ jsxRuntime.jsx(designSystem.ModalBody, { children: /* @__PURE__ */ jsxRuntime.jsx(
            designSystem.TextInput,
            {
              label: formatMessage({
                id: "content-releases.modal.form.input.label.release-name",
                defaultMessage: "Name"
              }),
              name: "name",
              value: values.name,
              error: errors.name,
              onChange: handleChange,
              required: true
            }
          ) }),
          /* @__PURE__ */ jsxRuntime.jsx(
            designSystem.ModalFooter,
            {
              startActions: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Button, { onClick: handleClose, variant: "tertiary", name: "cancel", children: formatMessage({ id: "cancel", defaultMessage: "Cancel" }) }),
              endActions: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Button, { name: "submit", loading: isLoading, disabled: !values.name, type: "submit", children: formatMessage({
                id: "content-releases.modal.form.button.submit",
                defaultMessage: "Continue"
              }) })
            }
          )
        ] })
      }
    )
  ] });
};
const ReleaseInfoWrapper = styled__default.default(designSystem.Flex)`
  align-self: stretch;
  border-bottom-right-radius: ${({ theme }) => theme.borderRadius};
  border-bottom-left-radius: ${({ theme }) => theme.borderRadius};
  border-top: 1px solid ${({ theme }) => theme.colors.neutral150};
`;
const StyledFlex = styled__default.default(designSystem.Flex)`
  align-self: stretch;
  cursor: ${({ disabled }) => disabled ? "not-allowed" : "pointer"};

  svg path {
    fill: ${({ theme, disabled }) => disabled && theme.colors.neutral500};
  }
  span {
    color: ${({ theme, disabled }) => disabled && theme.colors.neutral500};
  }
`;
const PencilIcon = styled__default.default(icons.Pencil)`
  width: ${({ theme }) => theme.spaces[4]};
  height: ${({ theme }) => theme.spaces[4]};
  path {
    fill: ${({ theme }) => theme.colors.neutral600};
  }
`;
const TrashIcon = styled__default.default(icons.Trash)`
  width: ${({ theme }) => theme.spaces[4]};
  height: ${({ theme }) => theme.spaces[4]};
  path {
    fill: ${({ theme }) => theme.colors.danger600};
  }
`;
const PopoverButton = ({ onClick, disabled, children }) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    StyledFlex,
    {
      paddingTop: 2,
      paddingBottom: 2,
      paddingLeft: 4,
      paddingRight: 4,
      alignItems: "center",
      gap: 2,
      as: "button",
      hasRadius: true,
      onClick,
      disabled,
      children
    }
  );
};
const ReleaseDetailsLayout = ({
  toggleEditReleaseModal,
  toggleWarningSubmit,
  children
}) => {
  const { formatMessage } = reactIntl.useIntl();
  const { releaseId } = reactRouterDom.useParams();
  const [isPopoverVisible, setIsPopoverVisible] = React__namespace.useState(false);
  const moreButtonRef = React__namespace.useRef(null);
  const {
    data,
    isLoading: isLoadingDetails,
    isError,
    error
  } = index.useGetReleaseQuery({ id: releaseId });
  const [publishRelease, { isLoading: isPublishing }] = index.usePublishReleaseMutation();
  const toggleNotification = helperPlugin.useNotification();
  const { formatAPIError } = helperPlugin.useAPIErrorHandler();
  const {
    allowedActions: { canUpdate, canDelete }
  } = helperPlugin.useRBAC(index.PERMISSIONS);
  const release = data?.data;
  const handleTogglePopover = () => {
    setIsPopoverVisible((prev) => !prev);
  };
  const openReleaseModal = () => {
    toggleEditReleaseModal();
    handleTogglePopover();
  };
  const handlePublishRelease = async () => {
    const response = await publishRelease({ id: releaseId });
    if ("data" in response) {
      toggleNotification({
        type: "success",
        message: formatMessage({
          id: "content-releases.pages.ReleaseDetails.publish-notification-success",
          defaultMessage: "Release was published successfully."
        })
      });
    } else if (index.isAxiosError(response.error)) {
      toggleNotification({
        type: "warning",
        message: formatAPIError(response.error)
      });
    } else {
      toggleNotification({
        type: "warning",
        message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
      });
    }
  };
  const openWarningConfirmDialog = () => {
    toggleWarningSubmit();
    handleTogglePopover();
  };
  if (isLoadingDetails) {
    return /* @__PURE__ */ jsxRuntime.jsx(designSystem.Main, { "aria-busy": isLoadingDetails, children: /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.LoadingIndicatorPage, {}) });
  }
  if (isError || !release) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      reactRouterDom.Redirect,
      {
        to: {
          pathname: "/plugins/content-releases",
          state: {
            errors: [
              {
                code: error?.code
              }
            ]
          }
        }
      }
    );
  }
  const totalEntries = release.actions.meta.count || 0;
  const createdBy = `${release.createdBy.firstname} ${release.createdBy.lastname}`;
  return /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Main, { "aria-busy": isLoadingDetails, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      designSystem.HeaderLayout,
      {
        title: release.name,
        subtitle: formatMessage(
          {
            id: "content-releases.pages.Details.header-subtitle",
            defaultMessage: "{number, plural, =0 {No entries} one {# entry} other {# entries}}"
          },
          { number: totalEntries }
        ),
        navigationAction: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Link, { startIcon: /* @__PURE__ */ jsxRuntime.jsx(icons.ArrowLeft, {}), to: "/plugins/content-releases", children: formatMessage({
          id: "global.back",
          defaultMessage: "Back"
        }) }),
        primaryAction: !release.releasedAt && /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { gap: 2, children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            designSystem.IconButton,
            {
              label: formatMessage({
                id: "content-releases.header.actions.open-release-actions",
                defaultMessage: "Release actions"
              }),
              ref: moreButtonRef,
              onClick: handleTogglePopover,
              children: /* @__PURE__ */ jsxRuntime.jsx(icons.More, {})
            }
          ),
          isPopoverVisible && /* @__PURE__ */ jsxRuntime.jsxs(
            designSystem.Popover,
            {
              source: moreButtonRef,
              placement: "bottom-end",
              onDismiss: handleTogglePopover,
              spacing: 4,
              minWidth: "242px",
              children: [
                /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { alignItems: "center", justifyContent: "center", direction: "column", padding: 1, children: [
                  /* @__PURE__ */ jsxRuntime.jsxs(PopoverButton, { disabled: !canUpdate, onClick: openReleaseModal, children: [
                    /* @__PURE__ */ jsxRuntime.jsx(PencilIcon, {}),
                    /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { ellipsis: true, children: formatMessage({
                      id: "content-releases.header.actions.edit",
                      defaultMessage: "Edit"
                    }) })
                  ] }),
                  /* @__PURE__ */ jsxRuntime.jsxs(PopoverButton, { disabled: !canDelete, onClick: openWarningConfirmDialog, children: [
                    /* @__PURE__ */ jsxRuntime.jsx(TrashIcon, {}),
                    /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { ellipsis: true, textColor: "danger600", children: formatMessage({
                      id: "content-releases.header.actions.delete",
                      defaultMessage: "Delete"
                    }) })
                  ] })
                ] }),
                /* @__PURE__ */ jsxRuntime.jsxs(
                  ReleaseInfoWrapper,
                  {
                    direction: "column",
                    justifyContent: "center",
                    alignItems: "flex-start",
                    gap: 1,
                    padding: 5,
                    children: [
                      /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { variant: "pi", fontWeight: "bold", children: formatMessage({
                        id: "content-releases.header.actions.created",
                        defaultMessage: "Created"
                      }) }),
                      /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Typography, { variant: "pi", color: "neutral300", children: [
                        /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.RelativeTime, { timestamp: new Date(release.createdAt) }),
                        formatMessage(
                          {
                            id: "content-releases.header.actions.created.description",
                            defaultMessage: " by {createdBy}"
                          },
                          { createdBy }
                        )
                      ] })
                    ]
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: index.PERMISSIONS.publish, children: /* @__PURE__ */ jsxRuntime.jsx(
            designSystem.Button,
            {
              size: "S",
              variant: "default",
              onClick: handlePublishRelease,
              loading: isPublishing,
              disabled: release.actions.meta.count === 0,
              children: formatMessage({
                id: "content-releases.header.actions.publish",
                defaultMessage: "Publish"
              })
            }
          ) })
        ] })
      }
    ),
    children
  ] });
};
const ReleaseDetailsBody = () => {
  const { formatMessage } = reactIntl.useIntl();
  const { releaseId } = reactRouterDom.useParams();
  const [{ query }] = helperPlugin.useQueryParams();
  const toggleNotification = helperPlugin.useNotification();
  const { formatAPIError } = helperPlugin.useAPIErrorHandler();
  const {
    data: releaseData,
    isLoading: isReleaseLoading,
    isError: isReleaseError,
    error: releaseError
  } = index.useGetReleaseQuery({ id: releaseId });
  const release = releaseData?.data;
  const {
    isLoading,
    isFetching,
    isError,
    data,
    error: releaseActionsError
  } = index.useGetReleaseActionsQuery({
    ...query,
    releaseId
  });
  const [updateReleaseAction] = index.useUpdateReleaseActionMutation();
  const handleChangeType = async (e, actionId) => {
    const response = await updateReleaseAction({
      params: {
        releaseId,
        actionId
      },
      body: {
        type: e.target.value
      }
    });
    if ("error" in response) {
      if (index.isAxiosError(response.error)) {
        toggleNotification({
          type: "warning",
          message: formatAPIError(response.error)
        });
      } else {
        toggleNotification({
          type: "warning",
          message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
        });
      }
    }
  };
  if (isLoading || isReleaseLoading) {
    return /* @__PURE__ */ jsxRuntime.jsx(designSystem.ContentLayout, { children: /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.LoadingIndicatorPage, {}) });
  }
  if (isError || isReleaseError || !release) {
    const errorsArray = [];
    if (releaseError) {
      errorsArray.push({
        code: releaseError.code
      });
    }
    if (releaseActionsError) {
      errorsArray.push({
        code: releaseActionsError.code
      });
    }
    return /* @__PURE__ */ jsxRuntime.jsx(
      reactRouterDom.Redirect,
      {
        to: {
          pathname: "/plugins/content-releases",
          state: {
            errors: errorsArray
          }
        }
      }
    );
  }
  const releaseActions = data?.data;
  const releaseMeta = data?.meta;
  if (!releaseActions || !releaseActions.length) {
    return /* @__PURE__ */ jsxRuntime.jsx(designSystem.ContentLayout, { children: /* @__PURE__ */ jsxRuntime.jsx(
      helperPlugin.NoContent,
      {
        content: {
          id: "content-releases.pages.Details.tab.emptyEntries",
          defaultMessage: "This release is empty. Open the Content Manager, select an entry and add it to the release."
        },
        action: /* @__PURE__ */ jsxRuntime.jsx(
          v2.LinkButton,
          {
            as: reactRouterDom.Link,
            to: {
              pathname: "/content-manager"
            },
            style: { textDecoration: "none" },
            variant: "secondary",
            children: formatMessage({
              id: "content-releases.page.Details.button.openContentManager",
              defaultMessage: "Open the Content Manager"
            })
          }
        )
      }
    ) });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(designSystem.ContentLayout, { children: /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { gap: 4, direction: "column", alignItems: "stretch", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      helperPlugin.Table.Root,
      {
        rows: releaseActions.map((item) => ({
          ...item,
          id: Number(item.entry.id)
        })),
        colCount: releaseActions.length,
        isLoading,
        isFetching,
        children: /* @__PURE__ */ jsxRuntime.jsxs(helperPlugin.Table.Content, { children: [
          /* @__PURE__ */ jsxRuntime.jsxs(helperPlugin.Table.Head, { children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              helperPlugin.Table.HeaderCell,
              {
                fieldSchemaType: "string",
                label: formatMessage({
                  id: "content-releases.page.ReleaseDetails.table.header.label.name",
                  defaultMessage: "name"
                }),
                name: "name"
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              helperPlugin.Table.HeaderCell,
              {
                fieldSchemaType: "string",
                label: formatMessage({
                  id: "content-releases.page.ReleaseDetails.table.header.label.locale",
                  defaultMessage: "locale"
                }),
                name: "locale"
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              helperPlugin.Table.HeaderCell,
              {
                fieldSchemaType: "string",
                label: formatMessage({
                  id: "content-releases.page.ReleaseDetails.table.header.label.content-type",
                  defaultMessage: "content-type"
                }),
                name: "content-type"
              }
            ),
            /* @__PURE__ */ jsxRuntime.jsx(
              helperPlugin.Table.HeaderCell,
              {
                fieldSchemaType: "string",
                label: formatMessage({
                  id: "content-releases.page.ReleaseDetails.table.header.label.action",
                  defaultMessage: "action"
                }),
                name: "action"
              }
            )
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.Table.LoadingBody, {}),
          /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.Table.Body, { children: releaseActions.map(({ id, type, entry }) => /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Tr, { children: [
            /* @__PURE__ */ jsxRuntime.jsx(designSystem.Td, { children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { children: `${entry.contentType.mainFieldValue || entry.id}` }) }),
            /* @__PURE__ */ jsxRuntime.jsx(designSystem.Td, { children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { children: `${entry?.locale?.name ? entry.locale.name : "-"}` }) }),
            /* @__PURE__ */ jsxRuntime.jsx(designSystem.Td, { children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { children: entry.contentType.displayName || "" }) }),
            /* @__PURE__ */ jsxRuntime.jsx(designSystem.Td, { children: release.releasedAt ? /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { children: formatMessage(
              {
                id: "content-releases.page.ReleaseDetails.table.action-published",
                defaultMessage: "This entry was <b>{isPublish, select, true {published} other {unpublished}}</b>."
              },
              {
                isPublish: type === "publish",
                b: (children) => /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { fontWeight: "bold", children })
              }
            ) }) : /* @__PURE__ */ jsxRuntime.jsx(
              index.ReleaseActionOptions,
              {
                selected: type,
                handleChange: (e) => handleChangeType(e, id),
                name: `release-action-${id}-type`
              }
            ) })
          ] }, id)) })
        ] })
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { paddingTop: 4, alignItems: "flex-end", justifyContent: "space-between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.PageSizeURLQuery, { defaultValue: releaseMeta?.pagination?.pageSize.toString() }),
      /* @__PURE__ */ jsxRuntime.jsx(
        helperPlugin.PaginationURLQuery,
        {
          pagination: {
            pageCount: releaseMeta?.pagination?.pageCount || 0
          }
        }
      )
    ] })
  ] }) });
};
const ReleaseDetailsPage = () => {
  const { formatMessage } = reactIntl.useIntl();
  const { releaseId } = reactRouterDom.useParams();
  const toggleNotification = helperPlugin.useNotification();
  const { formatAPIError } = helperPlugin.useAPIErrorHandler();
  const { push } = reactRouterDom.useHistory();
  const [releaseModalShown, setReleaseModalShown] = React__namespace.useState(false);
  const [showWarningSubmit, setWarningSubmit] = React__namespace.useState(false);
  const {
    isLoading: isLoadingDetails,
    data,
    isSuccess: isSuccessDetails
  } = index.useGetReleaseQuery({ id: releaseId });
  const [updateRelease, { isLoading: isSubmittingForm }] = index.useUpdateReleaseMutation();
  const [deleteRelease, { isLoading: isDeletingRelease }] = index.useDeleteReleaseMutation();
  const toggleEditReleaseModal = () => {
    setReleaseModalShown((prev) => !prev);
  };
  const toggleWarningSubmit = () => setWarningSubmit((prevState) => !prevState);
  if (isLoadingDetails) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      ReleaseDetailsLayout,
      {
        toggleEditReleaseModal,
        toggleWarningSubmit,
        children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.ContentLayout, { children: /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.LoadingIndicatorPage, {}) })
      }
    );
  }
  const title = isSuccessDetails && data?.data?.name || "";
  const handleEditRelease = async (values) => {
    const response = await updateRelease({
      id: releaseId,
      name: values.name
    });
    if ("data" in response) {
      toggleNotification({
        type: "success",
        message: formatMessage({
          id: "content-releases.modal.release-updated-notification-success",
          defaultMessage: "Release updated."
        })
      });
    } else if (index.isAxiosError(response.error)) {
      toggleNotification({
        type: "warning",
        message: formatAPIError(response.error)
      });
    } else {
      toggleNotification({
        type: "warning",
        message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
      });
    }
    toggleEditReleaseModal();
  };
  const handleDeleteRelease = async () => {
    const response = await deleteRelease({
      id: releaseId
    });
    if ("data" in response) {
      push("/plugins/content-releases");
    } else if (index.isAxiosError(response.error)) {
      toggleNotification({
        type: "warning",
        message: formatAPIError(response.error)
      });
    } else {
      toggleNotification({
        type: "warning",
        message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
      });
    }
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(
    ReleaseDetailsLayout,
    {
      toggleEditReleaseModal,
      toggleWarningSubmit,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ReleaseDetailsBody, {}),
        releaseModalShown && /* @__PURE__ */ jsxRuntime.jsx(
          ReleaseModal,
          {
            handleClose: toggleEditReleaseModal,
            handleSubmit: handleEditRelease,
            isLoading: isLoadingDetails || isSubmittingForm,
            initialValues: { name: title || "" }
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          helperPlugin.ConfirmDialog,
          {
            bodyText: {
              id: "content-releases.dialog.confirmation-message",
              defaultMessage: "Are you sure you want to delete this release?"
            },
            isOpen: showWarningSubmit,
            isConfirmButtonLoading: isDeletingRelease,
            onToggleDialog: toggleWarningSubmit,
            onConfirm: handleDeleteRelease
          }
        )
      ]
    }
  );
};
const ProtectedReleaseDetailsPage = () => /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: index.PERMISSIONS.main, children: /* @__PURE__ */ jsxRuntime.jsx(ReleaseDetailsPage, {}) });
const ReleasesLayout = ({
  isLoading,
  totalReleases,
  onClickAddRelease,
  children
}) => {
  const { formatMessage } = reactIntl.useIntl();
  return /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Main, { "aria-busy": isLoading, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      designSystem.HeaderLayout,
      {
        title: formatMessage({
          id: "content-releases.pages.Releases.title",
          defaultMessage: "Releases"
        }),
        subtitle: !isLoading && formatMessage(
          {
            id: "content-releases.pages.Releases.header-subtitle",
            defaultMessage: "{number, plural, =0 {No releases} one {# release} other {# releases}}"
          },
          { number: totalReleases }
        ),
        primaryAction: /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: index.PERMISSIONS.create, children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Button, { startIcon: /* @__PURE__ */ jsxRuntime.jsx(icons.Plus, {}), onClick: onClickAddRelease, children: formatMessage({
          id: "content-releases.header.actions.add-release",
          defaultMessage: "New release"
        }) }) })
      }
    ),
    children
  ] });
};
const LinkCard = styled__default.default(v2.Link)`
  display: block;
`;
const ReleasesGrid = ({ sectionTitle, releases = [], isError = false }) => {
  const { formatMessage } = reactIntl.useIntl();
  if (isError) {
    return /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.AnErrorOccurred, {});
  }
  if (releases?.length === 0) {
    return /* @__PURE__ */ jsxRuntime.jsx(
      designSystem.EmptyStateLayout,
      {
        content: formatMessage(
          {
            id: "content-releases.page.Releases.tab.emptyEntries",
            defaultMessage: "No releases"
          },
          {
            target: sectionTitle
          }
        ),
        icon: /* @__PURE__ */ jsxRuntime.jsx(icons.EmptyDocuments, { width: "10rem" })
      }
    );
  }
  return /* @__PURE__ */ jsxRuntime.jsx(designSystem.Grid, { gap: 4, children: releases.map(({ id, name, actions }) => /* @__PURE__ */ jsxRuntime.jsx(designSystem.GridItem, { col: 3, s: 6, xs: 12, children: /* @__PURE__ */ jsxRuntime.jsx(LinkCard, { href: `content-releases/${id}`, isExternal: false, children: /* @__PURE__ */ jsxRuntime.jsxs(
    designSystem.Flex,
    {
      direction: "column",
      justifyContent: "space-between",
      padding: 4,
      hasRadius: true,
      background: "neutral0",
      shadow: "tableShadow",
      height: "100%",
      width: "100%",
      alignItems: "start",
      gap: 2,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { as: "h3", variant: "delta", fontWeight: "bold", children: name }),
        /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { variant: "pi", children: formatMessage(
          {
            id: "content-releases.page.Releases.release-item.entries",
            defaultMessage: "{number, plural, =0 {No entries} one {# entry} other {# entries}}"
          },
          { number: actions.meta.count }
        ) })
      ]
    }
  ) }) }, id)) });
};
const INITIAL_FORM_VALUES = {
  name: ""
};
const ReleasesPage = () => {
  const location = reactRouterDom.useLocation();
  const [releaseModalShown, setReleaseModalShown] = React__namespace.useState(false);
  const toggleNotification = helperPlugin.useNotification();
  const { formatMessage } = reactIntl.useIntl();
  const { push, replace } = reactRouterDom.useHistory();
  const { formatAPIError } = helperPlugin.useAPIErrorHandler();
  const [{ query }, setQuery] = helperPlugin.useQueryParams();
  const response = index.useGetReleasesQuery(query);
  const [createRelease, { isLoading: isSubmittingForm }] = index.useCreateReleaseMutation();
  const { isLoading, isSuccess, isError } = response;
  React__namespace.useEffect(() => {
    if (location?.state?.errors) {
      toggleNotification({
        type: "warning",
        title: formatMessage({
          id: "content-releases.pages.Releases.notification.error.title",
          defaultMessage: "Your request could not be processed."
        }),
        message: formatMessage({
          id: "content-releases.pages.Releases.notification.error.message",
          defaultMessage: "Please try again or open another release."
        })
      });
      replace({ state: null });
    }
  }, [formatMessage, location?.state?.errors, replace, toggleNotification]);
  const toggleAddReleaseModal = () => {
    setReleaseModalShown((prev) => !prev);
  };
  if (isLoading) {
    return /* @__PURE__ */ jsxRuntime.jsx(ReleasesLayout, { onClickAddRelease: toggleAddReleaseModal, isLoading: true, children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.ContentLayout, { children: /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.LoadingIndicatorPage, {}) }) });
  }
  const totalReleases = isSuccess && response.currentData?.meta?.pagination?.total || 0;
  const handleTabChange = (index2) => {
    setQuery({
      ...query,
      page: 1,
      pageSize: response?.currentData?.meta?.pagination?.pageSize || 16,
      filters: {
        releasedAt: {
          $notNull: index2 === 0 ? false : true
        }
      }
    });
  };
  const activeTab = response?.currentData?.meta?.activeTab || "pending";
  const handleAddRelease = async (values) => {
    const response2 = await createRelease({
      name: values.name
    });
    if ("data" in response2) {
      toggleNotification({
        type: "success",
        message: formatMessage({
          id: "content-releases.modal.release-created-notification-success",
          defaultMessage: "Release created."
        })
      });
      push(`/plugins/content-releases/${response2.data.data.id}`);
    } else if (index.isAxiosError(response2.error)) {
      toggleNotification({
        type: "warning",
        message: formatAPIError(response2.error)
      });
    } else {
      toggleNotification({
        type: "warning",
        message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
      });
    }
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(ReleasesLayout, { onClickAddRelease: toggleAddReleaseModal, totalReleases, children: [
    /* @__PURE__ */ jsxRuntime.jsx(designSystem.ContentLayout, { children: /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsxs(
        designSystem.TabGroup,
        {
          label: formatMessage({
            id: "content-releases.pages.Releases.tab-group.label",
            defaultMessage: "Releases list"
          }),
          variant: "simple",
          initialSelectedTabIndex: ["pending", "done"].indexOf(activeTab),
          onTabChange: handleTabChange,
          children: [
            /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Box, { paddingBottom: 8, children: [
              /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Tabs, { children: [
                /* @__PURE__ */ jsxRuntime.jsx(designSystem.Tab, { children: formatMessage({
                  id: "content-releases.pages.Releases.tab.pending",
                  defaultMessage: "Pending"
                }) }),
                /* @__PURE__ */ jsxRuntime.jsx(designSystem.Tab, { children: formatMessage({
                  id: "content-releases.pages.Releases.tab.done",
                  defaultMessage: "Done"
                }) })
              ] }),
              /* @__PURE__ */ jsxRuntime.jsx(designSystem.Divider, {})
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs(designSystem.TabPanels, { children: [
              /* @__PURE__ */ jsxRuntime.jsx(designSystem.TabPanel, { children: /* @__PURE__ */ jsxRuntime.jsx(
                ReleasesGrid,
                {
                  sectionTitle: "pending",
                  releases: response?.currentData?.data,
                  isError
                }
              ) }),
              /* @__PURE__ */ jsxRuntime.jsx(designSystem.TabPanel, { children: /* @__PURE__ */ jsxRuntime.jsx(
                ReleasesGrid,
                {
                  sectionTitle: "done",
                  releases: response?.currentData?.data,
                  isError
                }
              ) })
            ] })
          ]
        }
      ),
      totalReleases > 0 && /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { paddingTop: 4, alignItems: "flex-end", justifyContent: "space-between", children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          helperPlugin.PageSizeURLQuery,
          {
            options: ["8", "16", "32", "64"],
            defaultValue: response?.currentData?.meta?.pagination?.pageSize.toString()
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          helperPlugin.PaginationURLQuery,
          {
            pagination: {
              pageCount: response?.currentData?.meta?.pagination?.pageCount || 0
            }
          }
        )
      ] })
    ] }) }),
    releaseModalShown && /* @__PURE__ */ jsxRuntime.jsx(
      ReleaseModal,
      {
        handleClose: toggleAddReleaseModal,
        handleSubmit: handleAddRelease,
        isLoading: isSubmittingForm,
        initialValues: INITIAL_FORM_VALUES
      }
    )
  ] });
};
const ProtectedReleasesPage = () => /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: index.PERMISSIONS.main, children: /* @__PURE__ */ jsxRuntime.jsx(ReleasesPage, {}) });
const App = () => {
  return /* @__PURE__ */ jsxRuntime.jsxs(reactRouterDom.Switch, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(reactRouterDom.Route, { exact: true, path: `/plugins/${index.pluginId}`, component: ProtectedReleasesPage }),
    /* @__PURE__ */ jsxRuntime.jsx(
      reactRouterDom.Route,
      {
        exact: true,
        path: `/plugins/${index.pluginId}/:releaseId`,
        component: ProtectedReleaseDetailsPage
      }
    )
  ] });
};
exports.App = App;
