// @strapi/content-releases/dist/_chunks/index-66d129ac.js
"use strict";
const helperPlugin = require("@strapi/helper-plugin");
const icons = require("@strapi/icons");
const jsxRuntime = require("react/jsx-runtime");
const React = require("react");
const query = require("@reduxjs/toolkit/query");
const designSystem = require("@strapi/design-system");
const v2 = require("@strapi/design-system/v2");
const axios = require("axios");
const formik = require("formik");
const reactIntl = require("react-intl");
const reactRouterDom = require("react-router-dom");
const yup = require("yup");
const react = require("@reduxjs/toolkit/query/react");
const styled = require("styled-components");
const _interopDefault = (e) => e && e.__esModule ? e : { default: e };
function _interopNamespace(e) {
  if (e && e.__esModule)
    return e;
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const React__namespace = /* @__PURE__ */ _interopNamespace(React);
const yup__namespace = /* @__PURE__ */ _interopNamespace(yup);
const styled__default = /* @__PURE__ */ _interopDefault(styled);
const __variableDynamicImportRuntimeHelper = (glob, path) => {
  const v = glob[path];
  if (v) {
    return typeof v === "function" ? v() : Promise.resolve(v);
  }
  return new Promise((_, reject) => {
    (typeof queueMicrotask === "function" ? queueMicrotask : setTimeout)(reject.bind(null, new Error("Unknown variable dynamic import: " + path)));
  });
};
const PERMISSIONS = {
  main: [
    {
      action: "plugin::content-releases.read",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ],
  create: [
    {
      action: "plugin::content-releases.create",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ],
  update: [
    {
      action: "plugin::content-releases.update",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ],
  delete: [
    {
      action: "plugin::content-releases.delete",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ],
  createAction: [
    {
      action: "plugin::content-releases.create-action",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ],
  deleteAction: [
    {
      action: "plugin::content-releases.delete-action",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ],
  publish: [
    {
      action: "plugin::content-releases.publish",
      subject: null,
      id: "",
      actionParameters: {},
      properties: {},
      conditions: []
    }
  ]
};
const pluginId = "content-releases";
const axiosBaseQuery = async ({
  url,
  method,
  data,
  config
}) => {
  try {
    const { get, post, del, put } = helperPlugin.getFetchClient();
    if (method === "POST") {
      const result2 = await post(url, data, config);
      return { data: result2.data };
    }
    if (method === "DELETE") {
      const result2 = await del(url, config);
      return { data: result2.data };
    }
    if (method === "PUT") {
      const result2 = await put(url, data, config);
      return { data: result2.data };
    }
    const result = await get(url, config);
    return { data: result.data };
  } catch (error) {
    const err = error;
    return {
      error: {
        status: err.response?.status,
        code: err.code,
        response: {
          data: err.response?.data
        }
      }
    };
  }
};
const isAxiosError = (err) => {
  return typeof err === "object" && err !== null && "response" in err && typeof err.response === "object" && err.response !== null && "data" in err.response;
};
const releaseApi = react.createApi({
  reducerPath: pluginId,
  baseQuery: axiosBaseQuery,
  tagTypes: ["Release", "ReleaseAction"],
  endpoints: (build) => {
    return {
      getReleasesForEntry: build.query({
        query(params) {
          return {
            url: "/content-releases",
            method: "GET",
            config: {
              params
            }
          };
        },
        providesTags: (result) => result ? [
          ...result.data.map(({ id }) => ({ type: "Release", id })),
          { type: "Release", id: "LIST" }
        ] : []
      }),
      getReleases: build.query({
        query({ page, pageSize, filters } = {
          page: 1,
          pageSize: 16,
          filters: {
            releasedAt: {
              $notNull: false
            }
          }
        }) {
          return {
            url: "/content-releases",
            method: "GET",
            config: {
              params: {
                page: page || 1,
                pageSize: pageSize || 16,
                filters: filters || {
                  releasedAt: {
                    $notNull: false
                  }
                }
              }
            }
          };
        },
        transformResponse(response, meta, arg) {
          const releasedAtValue = arg?.filters?.releasedAt?.$notNull;
          const isActiveDoneTab = releasedAtValue === "true";
          const newResponse = {
            ...response,
            meta: {
              ...response.meta,
              activeTab: isActiveDoneTab ? "done" : "pending"
            }
          };
          return newResponse;
        },
        providesTags: (result) => result ? [
          ...result.data.map(({ id }) => ({ type: "Release", id })),
          { type: "Release", id: "LIST" }
        ] : [{ type: "Release", id: "LIST" }]
      }),
      getRelease: build.query({
        query({ id }) {
          return {
            url: `/content-releases/${id}`,
            method: "GET"
          };
        },
        providesTags: (result, error, arg) => [{ type: "Release", id: arg.id }]
      }),
      getReleaseActions: build.query({
        query({ releaseId, page, pageSize }) {
          return {
            url: `/content-releases/${releaseId}/actions`,
            method: "GET",
            config: {
              params: {
                page,
                pageSize
              }
            }
          };
        },
        providesTags: (result, error, arg) => result ? [
          ...result.data.map(({ id }) => ({ type: "ReleaseAction", id })),
          { type: "ReleaseAction", id: "LIST" }
        ] : [{ type: "ReleaseAction", id: "LIST" }]
      }),
      createRelease: build.mutation({
        query(data) {
          return {
            url: "/content-releases",
            method: "POST",
            data
          };
        },
        invalidatesTags: [{ type: "Release", id: "LIST" }]
      }),
      updateRelease: build.mutation({
        query({ id, ...data }) {
          return {
            url: `/content-releases/${id}`,
            method: "PUT",
            data
          };
        },
        invalidatesTags: (result, error, arg) => [{ type: "Release", id: arg.id }]
      }),
      createReleaseAction: build.mutation({
        query({ body, params }) {
          return {
            url: `/content-releases/${params.releaseId}/actions`,
            method: "POST",
            data: body
          };
        },
        invalidatesTags: [
          { type: "Release", id: "LIST" },
          { type: "ReleaseAction", id: "LIST" }
        ]
      }),
      updateReleaseAction: build.mutation({
        query({ body, params }) {
          return {
            url: `/content-releases/${params.releaseId}/actions/${params.actionId}`,
            method: "PUT",
            data: body
          };
        },
        invalidatesTags: (result, error, arg) => [
          { type: "ReleaseAction", id: arg.params.actionId }
        ]
      }),
      deleteReleaseAction: build.mutation({
        query({ params }) {
          return {
            url: `/content-releases/${params.releaseId}/actions/${params.actionId}`,
            method: "DELETE"
          };
        },
        invalidatesTags: [
          { type: "Release", id: "LIST" },
          { type: "ReleaseAction", id: "LIST" }
        ]
      }),
      publishRelease: build.mutation({
        query({ id }) {
          return {
            url: `/content-releases/${id}/publish`,
            method: "POST"
          };
        },
        invalidatesTags: (result, error, arg) => [{ type: "Release", id: arg.id }]
      }),
      deleteRelease: build.mutation({
        query({ id }) {
          return {
            url: `/content-releases/${id}`,
            method: "DELETE"
          };
        },
        invalidatesTags: (result, error, arg) => [{ type: "Release", id: arg.id }]
      })
    };
  }
});
const {
  useGetReleasesQuery,
  useGetReleasesForEntryQuery,
  useGetReleaseQuery,
  useGetReleaseActionsQuery,
  useCreateReleaseMutation,
  useCreateReleaseActionMutation,
  useUpdateReleaseMutation,
  useUpdateReleaseActionMutation,
  usePublishReleaseMutation,
  useDeleteReleaseActionMutation,
  useDeleteReleaseMutation
} = releaseApi;
const StyledMenuItem = styled__default.default(v2.Menu.Item)`
  &:hover {
    background: transparent;
  }

  svg {
    path {
      fill: ${({ theme }) => theme.colors.danger600};
    }
  }

  &:hover {
    svg {
      path {
        fill: ${({ theme }) => theme.colors.danger600};
      }
    }
  }
`;
const StyledCross = styled__default.default(icons.Cross)`
  padding: ${({ theme }) => theme.spaces[1]};
`;
const ReleaseActionMenu = ({ releaseId, actionId }) => {
  const { formatMessage } = reactIntl.useIntl();
  const toggleNotification = helperPlugin.useNotification();
  const { formatAPIError } = helperPlugin.useAPIErrorHandler();
  const [deleteReleaseAction] = useDeleteReleaseActionMutation();
  const handleDeleteAction = async () => {
    const response = await deleteReleaseAction({
      params: { releaseId, actionId }
    });
    if ("data" in response) {
      toggleNotification({
        type: "success",
        message: formatMessage({
          id: "content-releases.content-manager-edit-view.remove-from-release.notification.success",
          defaultMessage: "Entry removed from release"
        })
      });
      return;
    }
    if ("error" in response) {
      if (axios.isAxiosError(response.error)) {
        toggleNotification({
          type: "warning",
          message: formatAPIError(response.error)
        });
      } else {
        toggleNotification({
          type: "warning",
          message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
        });
      }
    }
  };
  return (
    // A user can access the dropdown if they have permissions to delete a release-action OR update a release
    /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: [...PERMISSIONS.deleteAction, ...PERMISSIONS.update], children: /* @__PURE__ */ jsxRuntime.jsxs(v2.Menu.Root, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        v2.Menu.Trigger,
        {
          as: designSystem.IconButton,
          paddingLeft: 2,
          paddingRight: 2,
          "aria-label": formatMessage({
            id: "content-releases.content-manager-edit-view.release-action-menu",
            defaultMessage: "Release action options"
          }),
          icon: /* @__PURE__ */ jsxRuntime.jsx(icons.More, {})
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(v2.Menu.Content, { top: 1, children: /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: PERMISSIONS.deleteAction, children: /* @__PURE__ */ jsxRuntime.jsx(StyledMenuItem, { color: "danger600", onSelect: handleDeleteAction, children: /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { gap: 2, children: [
        /* @__PURE__ */ jsxRuntime.jsx(StyledCross, {}),
        /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { variant: "omega", children: formatMessage({
          id: "content-releases.content-manager-edit-view.remove-from-release",
          defaultMessage: "Remove from release"
        }) })
      ] }) }) }) })
    ] }) })
  );
};
const getBorderLeftRadiusValue = (actionType) => {
  return actionType === "publish" ? 1 : 0;
};
const getBorderRightRadiusValue = (actionType) => {
  return actionType === "publish" ? 0 : 1;
};
const FieldWrapper = styled__default.default(designSystem.Field)`
  border-top-left-radius: ${({ actionType, theme }) => theme.spaces[getBorderLeftRadiusValue(actionType)]};
  border-bottom-left-radius: ${({ actionType, theme }) => theme.spaces[getBorderLeftRadiusValue(actionType)]};
  border-top-right-radius: ${({ actionType, theme }) => theme.spaces[getBorderRightRadiusValue(actionType)]};
  border-bottom-right-radius: ${({ actionType, theme }) => theme.spaces[getBorderRightRadiusValue(actionType)]};

  > label {
    color: inherit;
    padding: ${({ theme }) => `${theme.spaces[2]} ${theme.spaces[3]}`};
    text-align: center;
    vertical-align: middle;
    text-transform: capitalize;
  }

  &:active,
  &[data-checked='true'] {
    color: ${({ theme }) => theme.colors.primary700};
    background-color: ${({ theme }) => theme.colors.primary100};
    border-color: ${({ theme }) => theme.colors.primary700};
  }

  &[data-checked='false'] {
    border-left: ${({ actionType }) => actionType === "unpublish" && "none"};
    border-right: ${({ actionType }) => actionType === "publish" && "none"};
  }
`;
const ActionOption = ({ selected, actionType, handleChange, name }) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    FieldWrapper,
    {
      actionType,
      background: "primary0",
      borderColor: "neutral200",
      color: selected === actionType ? "primary600" : "neutral600",
      position: "relative",
      cursor: "pointer",
      "data-checked": selected === actionType,
      children: /* @__PURE__ */ jsxRuntime.jsxs(designSystem.FieldLabel, { htmlFor: `${name}-${actionType}`, children: [
        /* @__PURE__ */ jsxRuntime.jsx(designSystem.VisuallyHidden, { children: /* @__PURE__ */ jsxRuntime.jsx(
          designSystem.FieldInput,
          {
            type: "radio",
            id: `${name}-${actionType}`,
            name,
            checked: selected === actionType,
            onChange: handleChange,
            value: actionType
          }
        ) }),
        actionType
      ] })
    }
  );
};
const ReleaseActionOptions = ({ selected, handleChange, name }) => {
  return /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ActionOption,
      {
        actionType: "publish",
        selected,
        handleChange,
        name
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      ActionOption,
      {
        actionType: "unpublish",
        selected,
        handleChange,
        name
      }
    )
  ] });
};
const RELEASE_ACTION_FORM_SCHEMA = yup__namespace.object().shape({
  type: yup__namespace.string().oneOf(["publish", "unpublish"]).required(),
  releaseId: yup__namespace.string().required()
});
const INITIAL_VALUES = {
  type: "publish",
  releaseId: ""
};
const NoReleases = () => {
  const { formatMessage } = reactIntl.useIntl();
  return /* @__PURE__ */ jsxRuntime.jsx(
    helperPlugin.NoContent,
    {
      content: {
        id: "content-releases.content-manager-edit-view.add-to-release.no-releases-message",
        defaultMessage: "No available releases. Open the list of releases and create a new one from there."
      },
      action: /* @__PURE__ */ jsxRuntime.jsx(
        v2.LinkButton,
        {
          to: {
            pathname: "/plugins/content-releases"
          },
          as: reactRouterDom.Link,
          variant: "secondary",
          children: formatMessage({
            id: "content-releases.content-manager-edit-view.add-to-release.redirect-button",
            defaultMessage: "Open the list of releases"
          })
        }
      )
    }
  );
};
const AddActionToReleaseModal = ({
  handleClose,
  contentTypeUid,
  entryId
}) => {
  const { formatMessage } = reactIntl.useIntl();
  const toggleNotification = helperPlugin.useNotification();
  const { formatAPIError } = helperPlugin.useAPIErrorHandler();
  const response = useGetReleasesForEntryQuery({
    contentTypeUid,
    entryId,
    hasEntryAttached: false
  });
  const releases = response.data?.data;
  const [createReleaseAction, { isLoading }] = useCreateReleaseActionMutation();
  const handleSubmit = async (values) => {
    const releaseActionEntry = {
      contentType: contentTypeUid,
      id: entryId
    };
    const response2 = await createReleaseAction({
      body: { type: values.type, entry: releaseActionEntry },
      params: { releaseId: values.releaseId }
    });
    if ("data" in response2) {
      toggleNotification({
        type: "success",
        message: formatMessage({
          id: "content-releases.content-manager-edit-view.add-to-release.notification.success",
          defaultMessage: "Entry added to release"
        })
      });
      handleClose();
      return;
    }
    if ("error" in response2) {
      if (axios.isAxiosError(response2.error)) {
        toggleNotification({
          type: "warning",
          message: formatAPIError(response2.error)
        });
      } else {
        toggleNotification({
          type: "warning",
          message: formatMessage({ id: "notification.error", defaultMessage: "An error occurred" })
        });
      }
    }
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(designSystem.ModalLayout, { onClose: handleClose, labelledBy: "title", children: [
    /* @__PURE__ */ jsxRuntime.jsx(designSystem.ModalHeader, { children: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { id: "title", fontWeight: "bold", textColor: "neutral800", children: formatMessage({
      id: "content-releases.content-manager-edit-view.add-to-release",
      defaultMessage: "Add to release"
    }) }) }),
    /* @__PURE__ */ jsxRuntime.jsx(
      formik.Formik,
      {
        onSubmit: handleSubmit,
        validationSchema: RELEASE_ACTION_FORM_SCHEMA,
        initialValues: INITIAL_VALUES,
        children: ({ values, setFieldValue }) => {
          return /* @__PURE__ */ jsxRuntime.jsxs(formik.Form, { children: [
            releases?.length === 0 ? /* @__PURE__ */ jsxRuntime.jsx(NoReleases, {}) : /* @__PURE__ */ jsxRuntime.jsx(designSystem.ModalBody, { children: /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { direction: "column", alignItems: "stretch", gap: 2, children: [
              /* @__PURE__ */ jsxRuntime.jsx(designSystem.Box, { paddingBottom: 6, children: /* @__PURE__ */ jsxRuntime.jsx(
                designSystem.SingleSelect,
                {
                  required: true,
                  label: formatMessage({
                    id: "content-releases.content-manager-edit-view.add-to-release.select-label",
                    defaultMessage: "Select a release"
                  }),
                  placeholder: formatMessage({
                    id: "content-releases.content-manager-edit-view.add-to-release.select-placeholder",
                    defaultMessage: "Select"
                  }),
                  onChange: (value) => setFieldValue("releaseId", value),
                  value: values.releaseId,
                  children: releases?.map((release) => /* @__PURE__ */ jsxRuntime.jsx(designSystem.SingleSelectOption, { value: release.id, children: release.name }, release.id))
                }
              ) }),
              /* @__PURE__ */ jsxRuntime.jsx(designSystem.FieldLabel, { children: formatMessage({
                id: "content-releases.content-manager-edit-view.add-to-release.action-type-label",
                defaultMessage: "What do you want to do with this entry?"
              }) }),
              /* @__PURE__ */ jsxRuntime.jsx(
                ReleaseActionOptions,
                {
                  selected: values.type,
                  handleChange: (e) => setFieldValue("type", e.target.value),
                  name: "type"
                }
              )
            ] }) }),
            /* @__PURE__ */ jsxRuntime.jsx(
              designSystem.ModalFooter,
              {
                startActions: /* @__PURE__ */ jsxRuntime.jsx(designSystem.Button, { onClick: handleClose, variant: "tertiary", name: "cancel", children: formatMessage({
                  id: "content-releases.content-manager-edit-view.add-to-release.cancel-button",
                  defaultMessage: "Cancel"
                }) }),
                endActions: (
                  /**
                   * TODO: Ideally we would use isValid from Formik to disable the button, however currently it always returns true
                   * for yup.string().required(), even when the value is falsy (including empty string)
                   */
                  /* @__PURE__ */ jsxRuntime.jsx(designSystem.Button, { type: "submit", disabled: !values.releaseId, loading: isLoading, children: formatMessage({
                    id: "content-releases.content-manager-edit-view.add-to-release.continue-button",
                    defaultMessage: "Continue"
                  }) })
                )
              }
            )
          ] });
        }
      }
    )
  ] });
};
const CMReleasesContainer = () => {
  const [isModalOpen, setIsModalOpen] = React__namespace.useState(false);
  const { formatMessage } = reactIntl.useIntl();
  const {
    isCreatingEntry,
    allLayoutData: { contentType }
  } = helperPlugin.useCMEditViewDataManager();
  const params = reactRouterDom.useParams();
  const canFetch = params?.id != null && contentType?.uid != null;
  const fetchParams = canFetch ? {
    contentTypeUid: contentType.uid,
    entryId: params.id,
    hasEntryAttached: true
  } : query.skipToken;
  const response = useGetReleasesForEntryQuery(fetchParams);
  const releases = response.data?.data;
  if (!canFetch) {
    return null;
  }
  if (isCreatingEntry || !contentType?.options?.draftAndPublish) {
    return null;
  }
  const toggleModal = () => setIsModalOpen((prev) => !prev);
  const getReleaseColorVariant = (actionType, shade) => {
    if (actionType === "unpublish") {
      return `secondary${shade}`;
    }
    return `success${shade}`;
  };
  return /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: PERMISSIONS.main, children: /* @__PURE__ */ jsxRuntime.jsxs(
    designSystem.Box,
    {
      as: "aside",
      "aria-label": formatMessage({
        id: "content-releases.plugin.name",
        defaultMessage: "Releases"
      }),
      background: "neutral0",
      borderColor: "neutral150",
      hasRadius: true,
      padding: 4,
      shadow: "tableShadow",
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { direction: "column", alignItems: "stretch", gap: 3, children: [
          /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { variant: "sigma", textColor: "neutral600", textTransform: "uppercase", children: formatMessage({
            id: "content-releases.plugin.name",
            defaultMessage: "Releases"
          }) }),
          releases?.map((release) => {
            return /* @__PURE__ */ jsxRuntime.jsxs(
              designSystem.Flex,
              {
                direction: "column",
                alignItems: "start",
                borderWidth: "1px",
                borderStyle: "solid",
                borderColor: getReleaseColorVariant(release.action.type, "200"),
                overflow: "hidden",
                hasRadius: true,
                children: [
                  /* @__PURE__ */ jsxRuntime.jsx(
                    designSystem.Box,
                    {
                      paddingTop: 3,
                      paddingBottom: 3,
                      paddingLeft: 4,
                      paddingRight: 4,
                      background: getReleaseColorVariant(release.action.type, "100"),
                      width: "100%",
                      children: /* @__PURE__ */ jsxRuntime.jsx(
                        designSystem.Typography,
                        {
                          fontSize: 1,
                          variant: "pi",
                          textColor: getReleaseColorVariant(release.action.type, "600"),
                          children: formatMessage(
                            {
                              id: "content-releases.content-manager-edit-view.list-releases.title",
                              defaultMessage: "{isPublish, select, true {Will be published in} other {Will be unpublished in}}"
                            },
                            { isPublish: release.action.type === "publish" }
                          )
                        }
                      )
                    }
                  ),
                  /* @__PURE__ */ jsxRuntime.jsxs(designSystem.Flex, { padding: 4, direction: "column", gap: 3, width: "100%", alignItems: "flex-start", children: [
                    /* @__PURE__ */ jsxRuntime.jsx(designSystem.Typography, { fontSize: 2, fontWeight: "bold", variant: "omega", textColor: "neutral700", children: release.name }),
                    /* @__PURE__ */ jsxRuntime.jsx(ReleaseActionMenu, { releaseId: release.id, actionId: release.action.id })
                  ] })
                ]
              },
              release.id
            );
          }),
          /* @__PURE__ */ jsxRuntime.jsx(helperPlugin.CheckPermissions, { permissions: PERMISSIONS.createAction, children: /* @__PURE__ */ jsxRuntime.jsx(
            designSystem.Button,
            {
              justifyContent: "center",
              paddingLeft: 4,
              paddingRight: 4,
              color: "neutral700",
              variant: "tertiary",
              startIcon: /* @__PURE__ */ jsxRuntime.jsx(icons.Plus, {}),
              onClick: toggleModal,
              children: formatMessage({
                id: "content-releases.content-manager-edit-view.add-to-release",
                defaultMessage: "Add to release"
              })
            }
          ) })
        ] }),
        isModalOpen && /* @__PURE__ */ jsxRuntime.jsx(
          AddActionToReleaseModal,
          {
            handleClose: toggleModal,
            contentTypeUid: contentType.uid,
            entryId: params.id
          }
        )
      ]
    }
  ) });
};
const admin = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  register(app) {
    if (window.strapi.features.isEnabled("cms-content-releases") && window.strapi.future.isEnabled("contentReleases")) {
      app.addMenuLink({
        to: `/plugins/${pluginId}`,
        icon: icons.PaperPlane,
        intlLabel: {
          id: `${pluginId}.plugin.name`,
          defaultMessage: "Releases"
        },
        async Component() {
          const { App } = await Promise.resolve().then(() => require("./App-f2cafd81.js"));
          return App;
        },
        permissions: PERMISSIONS.main
      });
      app.addMiddlewares([() => releaseApi.middleware]);
      app.addReducers({
        [releaseApi.reducerPath]: releaseApi.reducer
      });
      app.injectContentManagerComponent("editView", "right-links", {
        name: `${pluginId}-link`,
        Component: CMReleasesContainer
      });
    }
  },
  async registerTrads({ locales }) {
    const importedTrads = await Promise.all(
      locales.map((locale) => {
        return __variableDynamicImportRuntimeHelper(/* @__PURE__ */ Object.assign({ "./translations/en.json": () => Promise.resolve().then(() => require("./en-13576ce2.js")) }), `./translations/${locale}.json`).then(({ default: data }) => {
          return {
            data: helperPlugin.prefixPluginTranslations(data, "content-releases"),
            locale
          };
        }).catch(() => {
          return {
            data: {},
            locale
          };
        });
      })
    );
    return Promise.resolve(importedTrads);
  }
};
exports.PERMISSIONS = PERMISSIONS;
exports.ReleaseActionOptions = ReleaseActionOptions;
exports.admin = admin;
exports.isAxiosError = isAxiosError;
exports.pluginId = pluginId;
exports.useCreateReleaseMutation = useCreateReleaseMutation;
exports.useDeleteReleaseMutation = useDeleteReleaseMutation;
exports.useGetReleaseActionsQuery = useGetReleaseActionsQuery;
exports.useGetReleaseQuery = useGetReleaseQuery;
exports.useGetReleasesQuery = useGetReleasesQuery;
exports.usePublishReleaseMutation = usePublishReleaseMutation;
exports.useUpdateReleaseActionMutation = useUpdateReleaseActionMutation;
exports.useUpdateReleaseMutation = useUpdateReleaseMutation;
